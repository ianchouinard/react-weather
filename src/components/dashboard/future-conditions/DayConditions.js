import React, { Component } from 'react'
import { ConditionIcon } from './../../ConditionIcon'
import { Transforms } from '../../../classes/Transforms'
import { Conditions } from '../Conditions'
import PropTypes from 'prop-types'

export class DayConditions extends Component {

  render() {
    const {currentConditions} = this.props
  
    return (
      <div className="dayConditions">
        <div className="date">{ Transforms.date(this.props.currentConditions.dt) }</div>
        
        <div className="conditionStatus">
          <ConditionIcon icon={currentConditions.weather[0] ? currentConditions.weather[0].icon : ''} />
        </div>

        <Conditions
          conditions={currentConditions}
          title={`Conditions for ${Transforms.date(this.props.currentConditions.dt)}`} />
      </div>
    )
  }

}

DayConditions.propTypes = {
  currentConditions: PropTypes.shape({
    dt: PropTypes.date,
    weather: PropTypes.array.isRequired
  })
}
