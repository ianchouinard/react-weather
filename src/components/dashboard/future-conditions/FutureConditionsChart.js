import React, { Component } from 'react';
import { Line } from 'react-chartjs-2';
import { Transforms } from '../../../classes/Transforms'
import { store } from './../../../store/store'
import PropTypes from 'prop-types'

export class FutureConditionsChart extends Component {

  state = {
    chartData: {}
  }

  createChartData = () => {
    const dates = []
    const temps = []

    this.props.futureConditions.forEach((conditon) => {
      dates.push(Transforms.date(conditon.dt))
      temps.push(conditon.main.temp)
    })

    const unit = store.getState().unitType
    const tempUnit = (unit === 'metric') ? 'c' : 'f'

    return {
      labels: dates,
      datasets: [
        {
          data: temps,
          label: `Temp (${tempUnit})`,
          borderColor: "#3FBF7F",
          fill: true,
          backgroundColor: "rgba(38, 124, 81, 0.05)"
        }
      ]
    }
  }

  render() {
    const chartData = this.createChartData()

    return (
      <section className="chartContainer">
        <div className="container">
          <h5>Temperature Over Time</h5>
          <div className="chartWrapper">
            <div className="chartWrapperInner">
              <Line data={chartData} height={70}/>
            </div>
          </div>
        </div>
      </section>
    )
  }

}

FutureConditionsChart.propTypes = {
  futureConditions: PropTypes.array.isRequired
}
