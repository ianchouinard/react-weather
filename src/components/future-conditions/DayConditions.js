import React, { Component } from 'react'
import { ConditionIcon } from './../ConditionIcon'
import { Transforms } from '../../classes/Transforms'
import { store } from './../../store/store'

export class DayConditions extends Component {

  render() {
    const {currentConditions} = this.props
    const unit = store.getState().unitType
    const tempUnit = (unit === 'metric') ? 'c' : 'f'
    const speedUnit = (unit === 'metric') ? 'kph' : 'mph'
  
    return (
      <div className="dayConditions">
        <div className="date">{ Transforms.date(this.props.currentConditions.dt) }</div>
        
        <div className="conditionStatus">
          <ConditionIcon icon={currentConditions.weather[0] ? currentConditions.weather[0].icon : ''} />
        </div>

        <div className="conditions">
          <h3>{ currentConditions.weather[0] ? currentConditions.weather[0].main : 'N/A' }</h3>

          <p>Current Temp</p>
          <h3>{ Math.round(currentConditions.main.temp) } <sup>&deg; { tempUnit }</sup></h3>

          <p>Wind Speed</p>
          <h4>{ currentConditions.wind.speed } <sup>{ speedUnit }</sup></h4>
        </div>
      </div>
    )
  }

}
