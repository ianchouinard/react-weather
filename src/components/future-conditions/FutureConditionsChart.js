import React, { Component } from 'react';
import { Line } from 'react-chartjs-2';
import { Transforms } from '../../classes/Transforms';

export class FutureConditionsChart extends Component {

  state = {
    chartData: {}
  }

  componentWillReceiveProps() {
    setTimeout(() => {
      this.createChartData()
    }, 0)
  }

  createChartData = () => {
    const dates = []
    const temps = []

    this.props.futureConditions.forEach((conditon) => {
      dates.push(Transforms.date(conditon.dt))
      temps.push(conditon.main.temp)
    });

    this.setState({
      chartData: {
        labels: dates,
        datasets: [
          {
            data: temps,
            label: 'Temp (f)',
            borderColor: "#3FBF7F",
            fill: true,
            backgroundColor: "rgba(38, 124, 81, 0.05)"
          }
        ]
      }
    })
  }

  render() {
    return (
      <section className="chartContainer">
        <div className="container">
          <h5>Temperature Over Time</h5>
          <div className="chartWrapper">
            <div className="chartWrapperInner">
              <Line data={this.state.chartData} height={70}/>
            </div>
          </div>
        </div>
      </section>
    )
  }

}
