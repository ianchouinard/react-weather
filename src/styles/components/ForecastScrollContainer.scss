.forecastScrollContainer {
  margin-top: 40px;
  overflow: hidden;

  .scrollTitle {
    padding-left: 15px;
    font-weight: 700;
  }

  .scrollContainer {
    transition: all 350ms;

    .dayConditonWrapper {
      float: left;
      padding: 0 15px;
    }

    &:after {
      content: '';
        display: block;
        clear: both;
      }
    }

    .scrollControls {
      padding-top: 20px;
      text-align: center;

      button {
        margin: 0 5px;
        background: #282828;
        box-shadow: none;
        outline: none;
        border: none;
        cursor: pointer;

        svg path {
          fill: #fff;
        }

        &:hover {
          background: $color_bg;
                
          svg path {
            fill: #282828;
          }
        }

        &:first-of-type {
          transform: rotate(180deg);
        }
      }
    }
}