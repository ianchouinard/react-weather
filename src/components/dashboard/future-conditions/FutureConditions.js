import React, { Component } from 'react'
import { DayConditions } from './DayConditions'
import PropTypes from 'prop-types'

export class FutureConditions extends Component {

  /**
   * Determine the % width for each item based on the number of items.
   */
  getWrapperWidth = () => {
    return (100 / this.props.conditions.length )
  }

  render() {
    return this.props.conditions.map((condition) => (
      <div key={condition.dt} className="dayConditonWrapper" style={{width: `${this.getWrapperWidth()}%`}}>
        <DayConditions key={condition.dt} currentConditions={condition} />
      </div>
    ))
  }

}

FutureConditions.propTypes = {
  conditions: PropTypes.array.isRequired
}
