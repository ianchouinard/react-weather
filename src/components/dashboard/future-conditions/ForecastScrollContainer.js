import React, { Component } from 'react'
import { FutureConditions } from './FutureConditions'
import { ReactComponent as Arrow } from './../../../img/arrow.svg'
import PropTypes from 'prop-types'

export class ForecastScrollContainer extends Component {

  state = {
    xAxisTransform: 0,
    itemsPerPane: 5
  }

  /**
   * On mount, get the window size to determine how many items to show
   * at once in the window. Re run logic on resize.
   */
  componentDidMount() {
    this.calculateItemsPerPane(window.innerWidth)

    window.onresize = () => {
      this.setState({xAxisTransform: 0})
      this.calculateItemsPerPane(window.innerWidth)
    }
  }

  /**
   * Determine how many items to show per pane at a given width
   * @param {number} windowWidth
   */
  calculateItemsPerPane = (windowWidth) => {
    if (windowWidth < 480) {
      this.setState({itemsPerPane: 2})
    } else if (windowWidth < 880) {
      this.setState({itemsPerPane: 3})
    } else {
      this.setState({itemsPerPane: 5})
    }
  }

  /**
   * Get the width for the scrollable container based on the # of items
   */
  getScrollContainerWidth = () => {
    return (this.props.conditions.length / this.state.itemsPerPane) * 100
  }

  /**
   * View later forecasts by adding a negative translateX to the scroll view.
   */
  slideRight = () => {
    const translateAmount = (100 / this.props.conditions.length )
    const max = 100 - (translateAmount * this.state.itemsPerPane)

    if (Math.abs(this.state.xAxisTransform) < max) {
      this.setState({xAxisTransform: (this.state.xAxisTransform - translateAmount)})
    }
  }

  /**
   * View pealier forecasts by adding a positive translateX to the scrollview
   */
  slideLeft = () => {
    const translateAmount = this.state.xAxisTransform + (100 / this.props.conditions.length)

    if (translateAmount <= 0) {
      this.setState({xAxisTransform: translateAmount})
    }
  }

  render() {
    return (
      <div className="forecastScrollContainer container">
        <h3 className="scrollTitle">Future Forecast</h3>
        <div className="scrollContainer" style={{ width: `${this.getScrollContainerWidth()}%`, transform: `translateX(${this.state.xAxisTransform}%)` }}>
          <FutureConditions conditions={this.props.conditions} />
        </div>

        <div className="scrollControls">
          <button title="Previous" onClick={this.slideLeft}><Arrow /></button>
          <button title="Next" onClick={this.slideRight}><Arrow /></button>
        </div>
      </div>
    )
  }

}

ForecastScrollContainer.propTypes = {
  conditions: PropTypes.array.isRequired
}
