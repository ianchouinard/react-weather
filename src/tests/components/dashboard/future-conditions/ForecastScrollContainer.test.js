import React from 'react';
import renderer from 'react-test-renderer';
import { ForecastScrollContainer } from './../../../../components/dashboard/future-conditions/ForecastScrollContainer'
import { mockCondition } from '../../../Mocks';

describe('ForecastScollContainer', () => {
  const component = renderer.create(
    <ForecastScrollContainer conditions={[mockCondition()]} />
  )

  it('should create the component', () => {
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('calculateItemsPerPane()', () => {
    it('should set how many items to display', () => {
      jest.spyOn(component.getInstance(), 'setState')

      component.getInstance().calculateItemsPerPane(400)
      expect(component.getInstance().setState).toHaveBeenCalledWith({itemsPerPane: 2})

      component.getInstance().calculateItemsPerPane(800)
      expect(component.getInstance().setState).toHaveBeenCalledWith({itemsPerPane: 3})

      component.getInstance().calculateItemsPerPane(900)
      expect(component.getInstance().setState).toHaveBeenCalledWith({itemsPerPane: 5})
    })
  })

  describe('getScrollContainerWidth()', () => {
    it('should set the total width of the scroll container', () => {
      const expectedValue = 
        (component.getInstance().props.conditions.length / component.getInstance().state.itemsPerPane) * 100

      const testValue = component.getInstance().getScrollContainerWidth()

      expect(testValue).toBe(expectedValue)
    })
  })

  // const getScrollValuesMock = () => {
  //   return {
  //     items: component.getInstance().props.conditions.length,
  //     perPane: component.getInstance().state.itemsPerPane,
  //     axis: component.getInstance().state.xAxisTransform
  //   }
  // }

  // describe('slideRight()', () => {
  //   it('should set the state properties used for the scroll translate values', () => {
  //     const preScroll = getScrollValuesMock()
  //     jest.spyOn(component.getInstance(), 'setState')

  //     // {xAxisTransform: (this.state.xAxisTransform - translateAmount)}
  //     component.getInstance().slideRight()
  //     expect(component.getInstance().setState).toHaveBeenCalledWith(
  //       {xAxisTransform: -100}
  //     )
  //   })
  // })

})
