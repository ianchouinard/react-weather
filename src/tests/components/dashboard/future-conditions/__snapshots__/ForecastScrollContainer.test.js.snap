// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastScollContainer should create the component 1`] = `
<div
  className="forecastScrollContainer container"
>
  <h3
    className="scrollTitle"
  >
    Future Forecast
  </h3>
  <div
    className="scrollContainer"
    style={
      Object {
        "transform": "translateX(0%)",
        "width": "20%",
      }
    }
  >
    <div
      className="dayConditonWrapper"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <div
        className="dayConditions"
      >
        <div
          className="date"
        >
          10.26 4 AM
        </div>
        <div
          className="conditionStatus"
        >
          <svg>
            status_sunny.svg
          </svg>
        </div>
        <div
          className="conditions"
        >
          <h3 />
          <p>
            Current Temp
          </p>
          <h3>
            32
            <sup>
              Â° 
              F
            </sup>
          </h3>
          <p>
            Wind Speed
          </p>
          <h4>
            0
             
            <sup>
              mph
            </sup>
          </h4>
        </div>
      </div>
    </div>
  </div>
  <div
    className="scrollControls"
  >
    <button
      onClick={[Function]}
      title="Previous"
    >
      <svg>
        arrow.svg
      </svg>
    </button>
    <button
      onClick={[Function]}
      title="Next"
    >
      <svg>
        arrow.svg
      </svg>
    </button>
  </div>
</div>
`;
